[cli] Add support for CLI extensions in expo cli

chrfalch opened the pull request:
# Why

Expo already have an existing extension point in the Expo CLI called "Dev tools plugins" (https://docs.expo.dev/debugging/devtools-plugins/), which is the ability for an Expo package (npm library) to expose a web app that can be opened and hosted by the CLI using the `shift + m` shortcut. A devtools plugin can communicate with running Expo apps through web sockets and there are multiple examples of plugins available, from expo-atlas to redux plugin.

A missing feature in the current devtools plugin extension API is the ability to run simple one-shot commands from the CLI using the same mechanism - but launching commands instead of a web app in the browser, but still with support communication with any running Expo app.

Use-cases for command-line based CLI-extension can be to list and test expo background tasks from the command line, query a running app for its current screen (url) in expo-router or perform tasks like checking for SDK updates through the CLI (and maybe even upgrade the current project).

# How

I've extended the devtools configuration element in the expo-modules.config.json files so that we can define commands for the CLI using a configuration like this:

```json
{
  "platforms": [ "devtools" ],
  "devtools": {
	  "cliExtensions": {
	    "description": "Expo: tools for Expo App development",
	    "commands": [
	      { "name": "outdated", "title": "Check for outdated Expo packages", "environments": ["cli", "mcp"] },
	      { "name": "upgrade", "title": "Upgrade to latest Expo packages", "environments": ["cli", "mcp"] },
	      { "name": "take_screenshot", "title": "Take screenshot of Expo app", "environments": ["cli", "mcp"] },
	      { "name": "doctor", "title": "Run the Expo Doctor diagnostics tool", "environments": ["cli", "mcp"] }
	    ],
	    "entryPoint": "cli/build/index.js"
    }
  }
}
```

Each command will be displayed in the CLI interactive commands menu (shift+m) and will be run as node commands.

- expo-modules-autolinking:
  - Added types for the cli extension configuration
- cli:
  - Added support for showing the cli extensions as an interactive action (shift + m) in the cli
  - Added support for looking up and adding executors for cli externsions in DevToolsPluginManager.ts

# Test Plan

Build and run CLI. 

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

chrfalch committed:
[cli] Add support for CLI extensions in expo cli

This commit adds support for an new extension to the expo cli. This extension works in a similar matter as the devtools extension and is described in the notion doc here:

https://www.notion.so/expo/EXPO-CLI-Extension-225e5b57352480a8a707eb9627ea3c26

- expo-modules-autolinking
  - Added types for the cli extension configuration
- cli
  - Added support for showing the cli extensions as an interactive action (shift + m) in the cli
  - Added support for looking up and adding executors for cli externsions in DevToolsPluginManager.ts

chrfalch commented:
* **#38091** <a href="https://app.graphite.dev/github/pr/expo/expo/38091?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/expo/expo/38091?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Kudo commented on packages/expo-modules-autolinking/src/types.ts:
```diff
@@ -311,9 +325,59 @@ export interface RawExpoModuleConfig {
+         * Optional array of disabled environments for the command. By default all commands are enabled on all environments.
+         * Environments can be 'cli' for the CLI or 'mcp' for the Model Context Protocol.
+         */
+        disabled?: ('cli' | 'mcp')[];
```
maybe remove this first that we can add mcp from a separate pr.
i was thinking maybe `environments: ('cli'| 'mcp')[]` would be clearer

chrfalch replied:
Good point! What I did was to rename it to `environments` and then type it as an array of either `mcp` or `cli` - that way we don't need a redundant declaration and can mark a command directly with either one or both.

Kudo commented on packages/expo-modules-autolinking/src/types.ts:
```diff
@@ -311,9 +325,59 @@ export interface RawExpoModuleConfig {
+      /**
+       * The main entry point for the module in the CLI.
+       */
+      main: string;
```
```suggestion
      entryPoint: string;
```

Kudo commented on packages/@expo/cli/src/start/server/DevToolsPluginManager.ts:
```diff
@@ -27,7 +55,61 @@ export default class DevToolsPluginManager {
+                  shell: true,
+                  env: { ...process.env, FORCE_COLOR: '1' },
+                }
+              );
```
not sure whether we can show a hint in cli menu that menu will execute script from `tool` (e.g. `/app/node_modules/pkg/cli.js`)

chrfalch replied:
This is so that if the tool outputs data with formatted contents to stdout it will use this formatting info.

Kudo reviewed (changes requested):
as mentioned offline. i think it's good to introduce `zod` to validate the parameters passing from autolinking. becase we will then evaluate/execute command based on inputs

Kudo commented on packages/@expo/cli/src/start/server/DevToolsPluginManager.ts:
```diff
@@ -27,7 +55,61 @@ export default class DevToolsPluginManager {
+              const tool = path.join(plugin.packageRoot, plugin.cli!.main);
+              const child = spawn(
+                'node',
+                [tool, command, `'${JSON.stringify(args)}'`, `'${JSON.stringify(apps)}'`],
```
the payload from `apps` is long from command line execution. any chance we can prevent passing the apps?

also does `args` be used now? unclear to me where to specify the args. if the args are coming from expo-module.config.json, then why does the module need to pass args from expo-module.config.json -> autolinking -> cli -> node execution -> back to the module?

chrfalch replied:
**Apps**: We might change this to the server url for listing the connected devices, and then let the cli extension wrapper resolve it. I'll make a fix to this.
**Args**: coming from the MCP-server or the user (prompted for args). Typically the search query when performing a search

chrfalch commented on packages/@expo/cli/src/start/server/DevToolsPluginManager.ts:
```diff
@@ -27,7 +55,61 @@ export default class DevToolsPluginManager {
+      decsription: plugin.cli?.description ?? '',
+      executor: plugin.cli?.main
+        ? async ({ command, args, apps }: DevToolsPluginCliExecutorArguments) => {
+            return new Promise<string>(async (resolve, reject) => {
```
Seems to be available in @cli/packages.json?

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/h6M1FaHmn7sVoFKsGYxY/da702914-e271-4f6e-b92f-56d6b7714f2e.png)

chrfalch commented on packages/@expo/cli/src/start/interface/interactiveActions.ts:
```diff
@@ -141,22 +143,69 @@ export class DevServerManagerActions {
+          }
+          return null;
+        })
+        .filter((menuItem) => menuItem !== undefined) as MoreToolMenuItem[];
```
DONE!

chrfalch committed:
codereview: fixing renamed file that I'm struggling a bit with

chrfalch committed:
codereview: refactor apps argument

The apps argument passed to a tool might be a bit big to pass as a stringified json on the command line.

This commit fixes this by changing the apps argument to be the metro server origin so that the extension can use the `/json/list` endpoint by itself to get a list of connected apps.

The metro server origin should be available for both cli extensions and also future mcp extensions.
